Documentation/realtimebattle.6
.TH "realtimebattle" 6 "2008-12-11" "1.0.9" "RealTimeBattle"

.SH NAME
RealTimeBattle \- a robot programming game


.SH SYNOPSIS
\f[B]realtimebattle\f[R]
[\f[B]\-n\f[R]|\f[B]\-c\f[R]|\f[B]\-d\f[R]|\f[B]\-D\f[R] \f[I]debug-level\f[R]]
[\f[B]\-g\f[R]]
[\f[B]\-o\f[R] \f[I]option\-file\f[R]]
[\f[B]\-l\f[R] \f[I]logfile\f[R]]
[\f[B]\-t\f[R] \f[I]tournament-file\f[R]]
[\f[B]\-s\f[R] \f[I]statistics-file\f[R]]
[\f[B]\-p\f[R] \f[I]port-number\f[R]]
.br
\f[B]realtimebattle \-\-replay \f[I]logfile\f[R]
[\f[B]\-p\f[R] \f[I]port-number\f[R]]
.br
\f[B]realtimebattle -h \f[R]|\f[B] \-\-help
.br
\f[B]realtimebattle -v \f[R]|\f[B] \-\-version

.SH DESCRIPTION
\f[B]RealTimeBattle\f[R] is a programming game, in which robots controlled by programs compete against each other. The goal is to destroy the enemies, using the radar to examine the environment and the cannon to shoot.

Even though the environment the robots are moving in is fairly simple, it is far from easy to construct an intelligent robot\-program. \f[B]RealTimeBattle\f[R] is constructed to be easy to use, flexible and fast. The intention is that the program could be used for learning programming languages, as a test for intelligent algorithms or just as a game to play and enjoy.

Here is a list of features of \f[B]RealTimeBattle\f[R]:

.IP \(bu 4 
Game progresses in real time, with the robot programs running as child processes to \f[B]RealTimeBattle\f[R].
.IP \(bu 4 
The robots communicate with the main program using the standard input and output.
.IP \(bu 4 
Robots can be constructed in almost any programming language.
.IP \(bu 4 
Any number of robots can compete simultaneously if allowed by operating system and hardware.
.IP \(bu 4 
A simple messaging language is used for communication, which makes it easy to start constructing robots.
.IP \(bu 4 
Robots behave like a real physical object.
.IP \(bu 4 
Own arenas can be created.
.IP \(bu 4 
The simulation environment is highly configurable.
.IP \(bu 4 
It is possible to connect external clients to the simulation server.
.IP \(bu 4 
Basic team support is directly integrated in the game, sophisticated team play available due to team frameworks.
.IP \(bu 4 
\f[B]RealTimeBattle\f[R] is distributed under the General Public License (GPL). 
    

.SH OPTIONS

.TP
\fB\-\-debug_mode, \-d
debug mode

.TP
\fB\-\-debug_level \fI0-5\fB, \-D \fI0-5\fB
sets the initial debug level. implies \-d

.TP
\fB\-\-normal_mode, \-n
normal mode (default)

.TP
\fB\-\-competition_mode, \-c
competition mode

.TP
\fB\-\-no_graphics, \-g
no graphics will be displayed

.TP
\fB\-\-option_file \fIfile\fB, \-o \fIfile\fB
selects option\-file (default: $HOME/.rtbrc)

.TP
\fB\-\-log_file \fIfile\fB, \-l \fIfile\fB
make log file, if 'file' is '\-'
the log is sent to STDOUT

.TP
\fB\-\-tournament_file \fIfile\fB, \-t \fIfile\fB
specify a tournament file to autostart a tournament

.TP
\fB\-\-statistics_file \fIfile\fB, \-s \fIfile\fB
file to print the statistics to when autostarting

.TP
\fB\-\-message_file \fIfile\fB, \-m \fIfile\fB
redirect messages to 'file'.
'\-' as 'file' is equivalent to STDOUT.
If both log and messages are send
to STDOUT, '\-m' will be ignored

.TP
\fB\-\-replay \fIfile\fB, \-r \fIfile\fB
a log file to replay.
if '\-' is specified as file,
input is taken from STDIN

.TP
\fB\-\-help, \-h
prints this message

.TP
\fB\-\-version, \-v
prints the version number

.TP
\fB\-\-port_number, \-p
specifies the port for remote clients (default is 32134) (only available if compiled with --enable-network) 


.SH EXAMPLES

.TP
\fBrealtimebattle
Just start \f[B]RealTimeBattle\f[R] in normal mode with graphical user interface and listening for remote clients on port 32134 (if compiled with --enable-network).

.TP
\fBrealtimebattle \-D \fI5\fB
Enable full debugging output. Note that this is also useful for debugging robots, because the debug level is sent to the robots as a game option.

.TP
\fBrealtimebattle \-c \-g \-o \f[I]option\-file\f[B] -t \f[I]tournament\-file\f[B] -l \f[I]logfile\f[B] -s \f[I]statistics\-file\f[B]
A typical competition start of \f[B]RealTimeBattle\f[R]. There will be no graphical user interface and the results of the competion will be written to \f[I]logfile\f[R] and \f[I]statistics\-file\f[R].


.SH FILES
\f[I]/usr/local/games/RealTimeBattle/Documentation
.br
/usr/local/games/RealTimeBattle/Arenas
.br
/usr/local/games/RealTimeBattle/Robots
.br
/usr/local/include/RealTimeBattle/Messagetypes.h


.SH SEE ALSO
The full documentation for \f[B]RealTimeBattle\f[R] is maintained as an SGML manual and should have been installed together with the program in various formats.
You may have a look at the project website, which also contains the documentation: 
\f[I]http://realtimebattle.sourceforge.net/


.SH MAINTAINER
The current maintainer of the \f[B]RealTimeBattle\f[R] project is Johannes Nicolai <jonico@users.sourceforge.net>.
Feel free to contact him and the other developers at the source forge project page: \f[I]http://sourceforge.net/projects/realtimebattle/


.SH BUGS
If you find anything in this package which doesn't work, appears strange, is missing, is spellt wrong or is just confusing,
don't hesitate to send in a bug report to the Sourceforge Bug Tracker at: \f[I]http://sourceforge.net/tracker/?group_id=561

.SH AUTHOR
This manual page was written by Falko Menge <fakko@users.sourceforge.net> and is part of \f[B]RealTimeBattle\f[R].


ChangeLog
2005-09-19 Volker Stroebel <volker@planetpenguin.de>
	* configure.in include/Makefile.am include/BroadCast.h 
	include/ClientInterface.h src/Makefile.am src/Broadcast.cpp
	src/ClientInterface.cpp src/ArenaBase.cpp src/ArenaRealTime.cpp
	src/RealTimeBattle.cpp
	Added BroadCast class for new logging system
	Added ClientInterface class for experimental network support
	Changed ArenaRealTime to use BroadCast singleton for logfiles 

2004-11-12 Johannes Nicolai <jonico@users.sourceforge.net>
	* Created working autoconf / automake / autoheader structure
	* Imported some configuration data out of older versions and modified
	them so thatyou can install RealTime Battle on a modern system


src/RealTimeBattle.cc
/*
RealTimeBattle, a robot programming game for Unix
Copyright (C) 1998-2000  Erik Ouchterlony and Ragnar Ouchterlony

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software Foundation,
Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include <unistd.h>
#include <signal.h>

#include <sys/types.h>
#ifdef TIME_WITH_SYS_TIME 
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

# include <sys/wait.h>


#include <errno.h>
#ifdef HAVE_GETOPT_H
# include <getopt.h>
#else
# include "getopt.h"
#endif

#ifdef HAVE_LOCALE_H
# include <locale.h>
#endif

#include <iostream>

#ifdef FLOATINGPOINT_H
#include <floatingpoint>
#endif

#ifndef NO_GRAPHICS
#include "ControlWindow.h"
#endif 

#ifndef NO_NETWORK
#include "ClientInterface.h"
#endif

#include "ArenaRealTime.h"
#include "IntlDefs.h"
#include "Options.h"
#include "ArenaReplay.h"
#include "ArenaController.h"
#include <string>
#include <sstream>
#include "Various.h"
//#include "Server.h"

#ifndef WAIT_ANY
#define WAIT_ANY (pid_t)-1
#endif

class Options the_opts;
//class ArenaRealTime the_arena;
class ArenaController the_arena_controller;

//class Server my_server;

#ifndef NO_GRAPHICS
class Gui the_gui;
class ControlWindow* controlwindow_p;
#endif

bool no_graphics;



void 
print_help_message()
{
  cout << endl;
  cout << _(" Usage: RealTimeBattle [options] ") << endl << endl;
  cout << _(" Options:") << endl;
  cout << _("    --debug_mode,                -d   debug mode") << endl;
  cout << _("    --debug_level [0-5],         -D   sets the initial debug level. Implies -d") << endl;
  cout << _("    --normal_mode,               -n   normal mode (default)") << endl;
  cout << _("    --competition_mode,          -c   competition mode") << endl ;
  cout << endl;
  cout << _("    --no_graphics,               -g   no graphics will be displayed") << endl ;
  cout << _("    --option_file [file],        -o   selects option-file (default: $HOME/.rtbrc)")  << endl;
  cout << endl;
  cout << _("    --log_file [file],           -l   make log file, if 'file' is '-'\n"
            "                                      the log is sent to STDOUT") << endl;
  cout << endl;
  cout << _("    --tournament_file [file],    -t   specify a tournament file to\n"
            "                                      autostart a tournament") << endl;
  cout << endl;
  cout << _("    --statistics_file [file],    -s   file to print the statistics to\n"
            "                                       when autostarting") << endl;
  cout << endl;
  cout << _("    --message_file [file],       -m   redirect messages to 'file'.\n"
            "                                      With '-' as 'file', STDOUT is used.\n"
            "                                      If both log and messages are sent\n"
            "                                      to STDOUT, '-m' will be ignored") << endl;
  cout << endl;
  cout << _("    --replay [file]              -r   a log file to replay.\n"
            "                                      if '-' is specified as file,\n"
            "                                      input is taken from STDIN") << endl;
  cout << endl;
  cout << _("    --help,                      -h   prints this message") << endl;
  cout << _("    --version,                   -v   prints the version number") << endl;
  cout << endl;
}

#ifndef NO_GRAPHICS
gint
update_function(gpointer data)
{  
  gint res = 1;

  if( the_arena_controller.is_started() )
    res = (gint)the_arena.timeout_function();

  #ifndef NO_NETWORK
    ClientInterface::Instance()->loop();
  #endif
  
  return res;
}
#else 
void
update_function(const long int interval_usec)
{
  struct timeval timeout;
  timeout.tv_sec = 0;
  timeout.tv_usec = interval_usec;
  bool res = true;

  do
    {
      timeout.tv_sec = 0;
      timeout.tv_usec = interval_usec;
      select(FD_SETSIZE, NULL, NULL, NULL, &timeout);
      if( the_arena_controller.is_started() )
        res = the_arena.timeout_function();
	  
        #ifndef NO_NETWORK
        ClientInterface::Instance()->loop();
        #endif
    } 
  while( res );

}
#endif 

void
sig_handler (int signum)
{
  int pid;
  int status;
  while (1)
    {
      pid = waitpid (WAIT_ANY, &status, WNOHANG);
      if (pid < 0)
        {
          if( errno != ECHILD ) 
            Error(true, "waitpid failed", "RealTimeBattle.cc:sig_handler");
          break;
        }
      if (pid == 0)
        break;
    }
  //  if( signum == SIGCHLD ) cerr << "Sigchld caught!" << endl;
  //  if( signum == SIGPIPE ) cerr << "Sigpipe caught!" << endl;
  signal(signum, sig_handler);
}

void
sigfpe_handler(int signum)
{
  signal(signum, sigfpe_handler);
}

void
parse_command_line(int argc, char **argv)
{
  int version_flag=false, help_flag=false, graphics_flag=true;
  int c;
  istringstream string2number;

  extern char* optarg;
  extern int optind;

  static struct option long_options[] =
  {
    //option, argument?, flag, value
    {"version", 0, &version_flag, true},
    {"help", 0, &help_flag, true},

    {"debug_mode", 0, &the_arena_controller.game_mode,
     ArenaBase::DEBUG_MODE},
    {"debug_level", 1, 0, 0},
    {"normal_mode", 0, &the_arena_controller.game_mode,
     ArenaBase::NORMAL_MODE},
    {"competition_mode", 0, &the_arena_controller.game_mode,
     ArenaBase::COMPETITION_MODE},
    {"option_file", 1, 0, 0},
    {"log_file", 1, 0, 0},
    {"statistics_file", 1, 0, 0},
    {"tournament_file", 1, 0, 0},
    {"message_file", 1, 0, 0},
    {"replay", 1, 0, 0},

    {"no_graphics", 0, &graphics_flag, false},

    {0, 0, 0, 0}
  };

  for(;;)
    {
      int option_index = 0;
     
      c = getopt_long( argc, argv, "dncD:vho:l:s:t:m:r:g",
                       long_options, &option_index );

      // Detect the end of the options.
      if (c == -1)
        break;
     
      switch (c)
        {

        case 0:
          // If this option sets a flag, do nothing else now.
          if( long_options[option_index].flag != 0 )
            break;
          
          switch( option_index )
            {
            case 3:
              {
                string2number.clear();
                string2number.str(optarg);
                int optarg_num;
                string2number >> optarg_num;
                the_arena_controller.debug_level = optarg_num;
                the_arena_controller.game_mode = ArenaBase::DEBUG_MODE;
              }
              break;
            case 6: 
              the_arena_controller.option_filename = string(optarg);
              break;
            case 7:
              the_arena_controller.log_filename = string(optarg);
              break;
            case 8:
              the_arena_controller.statistics_filename = string(optarg);
              break;
            case 9:
              the_arena_controller.tournament_filename = string(optarg);
              the_arena_controller.auto_start_and_end = true;
              break;
            case 10:
              the_arena_controller.message_filename = string(optarg);
              break;
            case 11:
              the_arena_controller.replay_filename = string(optarg);
              break;
            default:
              Error( true, "Bad error, this shouldn't happen",
                     "RealTimeBattle.cc:parse_command_line" );
              exit( EXIT_FAILURE );
            }
          break;


        case 'd':
          the_arena_controller.game_mode = ArenaBase::DEBUG_MODE;
          break;

        case 'D':
          {
            string2number.clear();
            string2number.str(optarg);
            int optarg_num;
            string2number >> optarg_num;
            the_arena_controller.debug_level = optarg_num;
            the_arena_controller.game_mode = ArenaBase::DEBUG_MODE;
          }
          break;

        case 'n':
          the_arena_controller.game_mode = ArenaBase::NORMAL_MODE;
          break;

        case 'c':
          the_arena_controller.game_mode = ArenaBase::COMPETITION_MODE;
          break;

        case 'v':
          version_flag = true;
          break;

        case 'h':
          help_flag = true;
          break;

        case 'o':
          the_arena_controller.option_filename = string(optarg);
          break;

        case 'l':
          the_arena_controller.log_filename = string(optarg);
          break;

        case 's':
          the_arena_controller.statistics_filename = string(optarg);
          break;

        case 't':
          the_arena_controller.tournament_filename = string(optarg);
          the_arena_controller.auto_start_and_end = true;
          break;

        case 'm':
          the_arena_controller.message_filename = string(optarg);
          break;

        case 'r':
          the_arena_controller.replay_filename = string(optarg);
          break;

        case 'g':
          graphics_flag = false;
          break;

        default:
          print_help_message();
          exit( EXIT_FAILURE );
        }
    }

  if( the_arena_controller.debug_level > max_debug_level )
    the_arena_controller.debug_level = 5;
  if( the_arena_controller.debug_level < 0 )
    the_arena_controller.debug_level = 0;

  if(optind != argc) 
    {
      print_help_message();
      exit( EXIT_FAILURE );
    }

  if( version_flag )
      cout << "RealTimeBattle " << VERSION << endl;

  if( help_flag )
    print_help_message();

  if( help_flag || version_flag ) exit( EXIT_SUCCESS );

  if( the_arena_controller.option_filename.empty() )
    {
      the_opts.get_options_from_rtbrc();
      the_arena_controller.option_filename = ".rtbrc";
    }
  else
    the_opts.read_options_file( the_arena_controller.option_filename,true );

 // the_arena_controller.auto_start_and_end
 // = ( ( the_arena_controller.tournament_filename != "" ) ||
 //     ( the_arena_controller.replay_filename == "-" ) );
 
  the_arena_controller.auto_start_and_end
    = ( !the_arena_controller.tournament_filename.empty() ); 


  no_graphics = !graphics_flag;

  if( the_arena_controller.auto_start_and_end == false 
#ifndef NO_GRAPHICS      
      && no_graphics
#endif      
      )
    {
      cerr << "No graphics and no tournament file gives me nothing to do. Leaving ..." << endl;
      exit( EXIT_SUCCESS );
    }
}


int 
main ( int argc, char* argv[] )
{

#ifdef FLOATINGPOINT_H
  fpsetmask ( ~ (FP_X_INV | FP_X_DZ | FP_X_IMP | FP_X_OFL | FP_X_UFL | FP_X_DNML) );
  //fpsetmask ( ~ (FP_X_INV | FP_X_DZ | FP_X_IMP) );
#endif

#ifndef NO_GRAPHICS

	
#ifdef ENABLE_NLS
  bindtextdomain ("RealTimeBattle", RTB_LOCALEDIR);
  // allow german umlauts
  bind_textdomain_codeset ("RealTimeBattle", "UTF-8");
  textdomain ("RealTimeBattle");
#endif

  gtk_set_locale ();

  gtk_init (&argc, &argv);
#endif 
/*
#ifdef HAVE_LOCALE_H
  setlocale( LC_MESSAGES, "" );
  setlocale( LC_NUMERIC, "POSIX" );
#endif
#ifdef RTB_LOCALEDIR
  bindtextdomain( "RealTimeBattle", RTB_LOCALEDIR );
#else
  bindtextdomain( "RealTimeBattle", "/usr/local/share/locale" );
#endif
  textdomain( "RealTimeBattle" );
*/
  parse_command_line(argc, argv);

  if( !the_arena_controller.tournament_filename.empty() )
    the_arena_controller.start_realtime_arena();
  else if( !the_arena_controller.replay_filename.empty() )
    the_arena_controller.start_replay_arena();

  signal(SIGCHLD, sig_handler);
  signal(SIGPIPE, sig_handler);
  signal(SIGFPE, sigfpe_handler);

#ifndef NO_GRAPHICS
  if( !no_graphics )
    {
      the_gui.set_colours();
      controlwindow_p = 
        new ControlWindow( -1, -1,
                           the_opts.get_l( OPTION_CONTROL_WINDOW_POS_X ),
                           the_opts.get_l( OPTION_CONTROL_WINDOW_POS_Y ) );
    }


  gint timeout_tag;     

  double interval = 1000.0*the_opts.get_d( OPTION_UPDATE_INTERVAL ) - 10.0; 
  
  //updating too fast?
  if ( interval < 40 ) interval = 40;


  timeout_tag = gtk_timeout_add( (unsigned int) interval,
                                 GtkFunction(update_function), (gpointer) NULL );
 
  gtk_main();
#else 
  
  update_function( (long int)(the_opts.get_d( OPTION_UPDATE_INTERVAL ) * 1000000.0) );
  
#endif 

  return EXIT_SUCCESS;
}
